sway.el

```diff
@@ -328,24 +328,30 @@ TREE, VISIBLE-ONLY, FOCUSED-ONLY and return value are as in
 ;;;; (Interactive) helpers
 
 (defun sway--completing-read-container (prompt &optional id-only)
-  "Prompt the user to pick a Sway container, and return its ID."
+  "PROMPT the user to pick a Sway container, and return its ID.
+
+If ID-ONLY is nil of undefined, return the hashtable corresponding to
+the container."
   (let* ((windows (sway-list-windows))
          ;; Alist of (formatted-window-name . hashtable)
          (alist (mapcar (lambda (it)
-                        (cons
-                         (format "%s (%s)"
-                                 (gethash "name" it)
-                                 (gethash "id" it))
-                         it)) windows))
-         (choice (alist-get
-                  (completing-read prompt (mapcar 'car alist)) alist nil nil 'string=)))
+                          (cons
+                           (format "%s (%s)"
+                                   (gethash "name" it)
+                                   (gethash "id" it))
+                           it)) windows))
+         (selected (completing-read
+                    prompt
+                    (lambda (string predicate action)
```

Comment:
This function basically plays the role of the marginalia package : label candidates as `sway-container` so that embark knows that it can use `sway-container-embark-actions` as actions on these items when prompted. 

Now for the in-depth understanding of how this works, this is borrowed code, but IIU&RC : 
- action is the embark action (the action is present in the embark-menu, focus or kill here, you can switch embark-actions several times). 
- before we have an available action (metadata action), just add the category `sway-container` as the category metadata. 

I have to go, will complete the co..ent later.

Reply:
The proper documentation is in the info node "Programmed completion" : https://www.gnu.org/software/emacs/manual/html_node/elisp/Programmed-Completion.html

- in the case of metadata:
  "This specifies a request for information about the state of
   the current completion." For `category` metadata :
  "The value should be a symbol describing what kind of text the
  completion function is trying to complete."

- in the case of `complete-with-action`:
  `complete-with-action` is run with action which is neither metadata, nil or t, so it will behave like `test-completion`.
  So it will indeed run the completion.

Reply:
Thanks for the resource. Please just add this as even a very short comment, like:

```elisp
;; See (info "(elisp) Programmed completion")
```

Reply:
Should be good now.